### ISSUE TITLE
Issue installing caracal on whiplash node

### ISSUE THREAD

**Issue by shilparanchod** [caracal-pipeline/caracal#1094]
I'm getting the following error when pulling stimela images on whiplash:

Traceback (most recent call last):
  File "/scratch/users/shilpa/caracal-venv/bin/stimela", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/scratch/users/shilpa/Stimela/bin/stimela", line 14, in <module>
    main.main([a for a in sys.argv[1:]])
  File "/scratch/users/shilpa/Stimela/stimela/main.py", line 592, in main
    _cmd(argv)
  File "/scratch/users/shilpa/Stimela/stimela/main.py", line 309, in pull
    image, simage, directory=pull_folder, force=args.force)
  File "/scratch/users/shilpa/Stimela/stimela/singularity.py", line 39, in pull
    os.mkdir(directory)
PermissionError: [Errno 13] Permission denied: '/singularity_images'

@KshitijT

**Comment by KshitijT** [comment]
Thanks for raising this issue, @shilparanchod .

I confirm I get the same error. This occurs with the version of Stimela installed by CARACal and the Stimela master branch.

Moreover, the same procedure was working yesterday, so I don't know what changed overnight.  @SpheMakh , @Jordatious any clues?

**Comment by o-smirnov** [comment]
Werid. Are you just running ``stimela pull`` on the command line and getting this failure? Or via caratekit or something? What is your ``SINGULARITY_PULLFOLDER`` variable set to, if anything?

**Comment by SpheMakh** [comment]
@shilparanchod @KshitijT can you post the command you used when you got this error?

**Comment by shilparanchod** [comment]
command is `stimela pull -s` and I have `export SINGULARITY_PULLFOLDER=${WORKSPACE_ROOT}/singularity_images`

**Comment by SpheMakh** [comment]
I have a feeling that `${WORKSPACE_ROOT}` is empty, and you do not have permissions to write to `/`

what do you get when you run 

```
echo $WORKSPACE_ROOT
```

**Comment by shilparanchod** [comment]
Yeah, looks like it is empty. Weird, because it was working for me yesterday

**Comment by KshitijT** [comment]
On the other hand, I have NOT specified singularity pullfolder variable and I get the same error. Also, it was working fine yesterday until I decided to be _smart_ and update.

**Comment by o-smirnov** [comment]
Ah yes, I had a feeling this way of documenting things was going to confuse people. Not everyone's brain works like @gigjozsa's. Maybe we should simplify the documentation. But when you see something like this:
```
SINGULARITY_PULLFOLDER=${WORKSPACE_ROOT}/singularity_images
```
...what it means is that you have your own "workspace" directory where you plan to run things (this could even just be your home, or some project directory you make under your home), and ``SINGULARITY_PULLFOLDER`` should be set to a directory under that. 

WORKSPACE_ROOT by itself has no particular meaning to Caracal or Stimela. It's merely convenient to have WORKSPACE_ROOT set to point to your workspace directory, because then you can just cut and paste commands like the above and all the directories are set relative to that. The disadvantage is that if you forget to set WORKSPACE_ROOT, your cut-and-paste results in setting paths relative to "/", which is not what you want at all.

Bottom line is, please set ``SINGULARITY_PULLFOLDER`` to a directory you have write access to, then rerun.

**Comment by o-smirnov** [comment]
> until I decided to be smart and update.

Tut-tut, an old hand like you, falling for the oldest trick in the book! Well what's **your** SINGULARITY_PULLFOLDER set to?

**Comment by KshitijT** [comment]
> Tut-tut, an old hand like you, falling for the oldest trick in the book! Well what's your SINGULARITY_PULLFOLDER set to?

It isn't, that's the funny thing. I used to make a folder, cd to it and stimela pull -s and specify path in CARACal run. Was working fine. :)

I got the notion somehow that if you don't specify it and run the pipeline, it defaults to the working folder - and that was working fine till yesterday. Today, nope.

**Comment by KshitijT** [comment]
> if you don't specify it and run the pipeline, it defaults to the working folder...

In the new - fangled Stimela way, of course.

**Comment by o-smirnov** [comment]
It's supposed to. So if you don't have an explicit pullfolder set and it tries to go to ``/``, that's a new bug.

**Comment by gigjozsa** [comment]
> Not everyone's brain works like @gigjozsa's.

And @paoloserra 's actually who had the last word here.

**Comment by KshitijT** [comment]
> > Not everyone's brain works like @gigjozsa's.
> 
> And @paoloserra 's actually who had the last word here.

Git blame?

**Comment by o-smirnov** [comment]
No blame, but let's add a note to the docs to clarify this?

**Comment by o-smirnov** [comment]
Sorry @gigjozsa, I first saw this way of referencing the workspace dir in the caratekit help, so I assumed it was your invention. :)

**Comment by gigjozsa** [comment]
It probably is, but does it not really really make sense? Apparently not. We should not use names like that but just like A B, I guess.

**Comment by KshitijT** [comment]
To recap, @shilparanchod , please set the correct path to your stimela pullfolder in the bashrc file and try re-pulling.  Could you please comment here if that works for you?

**Comment by KshitijT** [comment]
The error I get on the other hand is because of the damned .singularity folder, stored right in my home directory, where I have limited storage (and rightly so). @o-smirnov , do you know any solution to that other than nuking .singularity ?

**Comment by gigjozsa** [comment]
Symbolic link. Or use caratekit (which does the same).

**Comment by o-smirnov** [comment]
Yeah. I also recommend symlinking things like ``~/.docker`` and ``~/.local`` and ``~/.cache`` while you're at it, these things tend to sneakily grow and grow until you suddenly hit your quota and you don't know why...

**Comment by paoloserra** [comment]
Hey there, just catching up, please feel free to change our README as you see fit.

**Comment by o-smirnov** [comment]
> It probably is, but does it not really really make sense? Apparently not.

Well, personally, I think like a programmer, so things like 
```
 SINGULARITY_PULLFOLDER=${WORKSPACE_ROOT}/singularity_images
```
...make perfect sense to me, I see a "mental variable" that could also well be a shell variable. But I can also see that to a new user, it may convey the impression that WORKSPACE_ROOT has some special meaning to the pipeline, when actually it doesn't. So just a tiny clarification will eliminate the potential confusion.

**Comment by gigjozsa** [comment]
Maybe we should not use this notation.

**Comment by o-smirnov** [comment]
The notation is fine, we just need to add a sentence at the start explaining it.

**Comment by SpheMakh** [comment]
Let's just keep it simple. 

----
Add this line to your ~/.bashrc

```
export SINGULARITY_PULLFOLDER=<folder to save stimela singularity images>
```

or this line if you are using a cshell

```
blah blah
```

**Comment by KshitijT** [comment]
I already modified the line on GitHub, can someone take a look?

**Comment by shilparanchod** [comment]
Even with changing the line to ~/.bashrc, I still get the following:

(caracal-venv) shilpa@whiplash:~/singularity_pull_folder$ stimela pull -s

Traceback (most recent call last):
  File "/scratch/users/shilpa/caracal-venv/bin/stimela", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/scratch/users/shilpa/Stimela/bin/stimela", line 14, in <module>
    main.main([a for a in sys.argv[1:]])
  File "/scratch/users/shilpa/Stimela/stimela/main.py", line 592, in main
    _cmd(argv)
  File "/scratch/users/shilpa/Stimela/stimela/main.py", line 309, in pull
    image, simage, directory=pull_folder, force=args.force)
  File "/scratch/users/shilpa/Stimela/stimela/singularity.py", line 39, in pull
    os.mkdir(directory)
PermissionError: [Errno 13] Permission denied: '/singularity_pull_folder'

**Comment by KshitijT** [comment]
@shilparanchod , could you please post the line in the bashrc?

**Comment by shilparanchod** [comment]
`export SINGULARITY_PULLFOLDER=${WORKSPACE_ROOT}/singularity_pull_folder`

**Comment by KshitijT** [comment]
To confirm, 
`export SINGULARITY_PULLFOLDER=<folder to save stimela singularity images>` is also set?

**Comment by gigjozsa** [comment]
I just edited the README...

**Comment by paoloserra** [comment]
@KshitijT -- all -- please do not edit the caracal-docs files directly. You should edit the README of the caracal repository, and the caracal-docs will be updated accordingly with a script.

**Comment by KshitijT** [comment]
> @KshitijT -- all -- please do not edit the caracal-docs files directly. You should edit the README of the caracal repository, and the caracal-docs will be updated accordingly with a script.

Sorry about that, my bad.

**Comment by shilparanchod** [comment]
Okay, its working for me. Thanks, all!
